### PR TITLE
auth: implement frontend parts of iap-like sign-in

### PR DESCRIPTION
**Pull request by freben**
This introduces a `ProxiedSignInPage` into `@backstage/core-components`.

Its intended purpose is to be a reusable sign-in page to use for adopters that rely on a reverse authenticating proxy to provide auth. This sign-in page has no user interface of its own, but rather supplies an identity API implementation which keeps itself up to date based on the corresponding auth-backend provider's `refresh` endpoint.

This page is a good fit for AWS ALB and GCP IAP providers, and possibly others to come that share the same semantics of being passive and "refresh-only".

### CODE REVIEW

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getCredentials} */
+  async getCredentials(): Promise<{ token?: string | undefined }> {
+    return { token: this.getSession().backstageIdentity.token };
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.signOut} */
+  async signOut(): Promise<void> {
+    this.abortController.abort();
```
Comment (freben) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts:154]: I wasn't sure what to actually do here. Should we just throw an exception? Should we refresh the page? Should we for real try to sign out of the session somehow?

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+      previous,
+    };
+
+    return promise;
+  }
+
+  async fetchSession(): Promise<DelegatedSession> {
+    const baseUrl = await this.options.discoveryApi.getBaseUrl('auth');
+
+    const response = await this.options.fetchApi.fetch(
```
Comment (Rugvip) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts:188]: Reminder to switch this to plain `fetch`

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInPage.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * This sign-in page is useful when you are using products such as Google
+ * Identity-Aware Proxy or AWS Application Load Balancer or similar, to front
+ * your Backstage installation. This sign-in page implementation will silently
+ * and regularly punch through the proxy to the auth backend to refresh your
+ * frontend session information, without requiring user interaction.
+ *
+ * @public
+ */
+export const DelegatedSignInPage = (props: DelegatedSignInPageProps) => {
```
Comment (Rugvip) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInPage.tsx:57]: Suggesting a rename to `ProxiedSignInPage` or smth like that. Another option could be to mix it in with the existing `SignInPage`, something like `<SignInPage proxied provider='oauthproxy' />`

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+// When the token expires, with some margin
+export function tokenToExpiry(jwtToken: string | undefined): Date {
+  if (!jwtToken) {
+    return new Date(Date.now() + DEFAULTS.defaultTokenExpiryMillis);
+  }
+
+  const [_header, rawPayload, _signature] = jwtToken.split('.');
+  const payload = JSON.parse(atob(rawPayload));
+
+  return new Date(payload.exp * 1000 - DEFAULTS.tokenExpiryMarginMillis);
```
Comment (Rugvip) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts:45]: Perhaps do a `payload.exp` check here and fall back to default if not set for safety?

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  IdentityApi,
+  ProfileInfo,
+} from '@backstage/core-plugin-api';
+import { ResponseError } from '@backstage/errors';
+import { DelegatedSession, delegatedSessionSchema } from './types';
+
+export const DEFAULTS = {
+  // The amount of time between token refreshes, if we fail to get an actual
+  // value out of the exp claim
+  defaultTokenExpiryMillis: 60_000,
```
Comment (Rugvip) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts:30]: I think we could at least default it to 5-10 min or so, would be highly unusual to have a token that only lasts a minute

File: packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  // The amount of time before the actual expiry of the Backstage token, that we
+  // shall start trying to get a new one
+  tokenExpiryMarginMillis: 30_000,
```
Comment (Rugvip) [inline packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts:33]: I think 30s is a bit low margin for the long-lived tokens, for the other OAuth2 providers we use 5 minutes for 1h tokens.

**Review by Rugvip — COMMENTED** [review]
Nice 😁

**Comment by freben** [comment]
Aight, all comments addressed!

**Review by Rugvip — APPROVED** [review]
🎉 :shipit:

**Pull request merged**